tests/v2/validation/snapshot/snapshot.go

```diff
@@ -290,51 +293,69 @@ func snapshotRestoreRKE2K3S(t *testing.T, client *rancher.Client, podTemplate co
-		Name:             snapshotToRestore,
-		Generation:       clusterObject.Spec.RKEConfig.ETCDSnapshotCreate.Generation,
-		RestoreRKEConfig: etcdRestore.SnapshotRestore,
-	}
+	// Give the option to restore the same snapshot multiple times. By default, it is set to 1.
+	for i := 0; i < etcdRestore.SnapshotCounter; i++ {
+		generation := int(1)
+		if clusterObject.Spec.RKEConfig.ETCDSnapshotRestore != nil {
+			generation = clusterObject.Spec.RKEConfig.ETCDSnapshotRestore.Generation + 1
+		}
 
-	err = etcdsnapshot.RestoreRKE2K3SSnapshot(client, clusterName, snapshotRKE2K3SRestore)
-	require.NoError(t, err)
+		snapshotRKE2K3SRestore := &rkev1.ETCDSnapshotRestore{
+			Name:             snapshotToRestore,
+			Generation:       generation,
+			RestoreRKEConfig: etcdRestore.SnapshotRestore,
+		}
 
-	err = clusters.WaitClusterToBeUpgraded(client, clusterID)
-	require.NoError(t, err)
+		err = etcdsnapshot.RestoreRKE2K3SSnapshot(client, clusterName, snapshotRKE2K3SRestore)
+		require.NoError(t, err)
 
-	clusterObject, _, err = clusters.GetProvisioningClusterByName(client, clusterName, namespace)
-	require.NoError(t, err)
+		err = clusters.WaitClusterToBeUpgraded(client, clusterID)
+		require.NoError(t, err)
 
-	podErrors = pods.StatusPods(client, clusterID)
-	assert.Empty(t, podErrors)
-	require.Equal(t, initialKubernetesVersion, clusterObject.Spec.KubernetesVersion)
+		clusterObject, _, err = clusters.GetProvisioningClusterByName(client, clusterName, namespace)
+		require.NoError(t, err)
 
-	steveclient, err := client.Steve.ProxyDownstream(clusterID)
-	require.NoError(t, err)
+		logrus.Infof("Cluster version is restored to: %s", clusterObject.Spec.KubernetesVersion)
 
-	deploymentList, err := steveclient.SteveType(workloads.DeploymentSteveType).NamespacedSteveClient(defaultNamespace).List(nil)
-	require.NoError(t, err)
-	require.Equal(t, 1, len(deploymentList.Data))
-	require.Equal(t, initialWorkloadName, deploymentList.Data[0].ObjectMeta.Name)
+		podErrors = pods.StatusPods(client, clusterID)
+		assert.Empty(t, podErrors)
+		require.Equal(t, initialKubernetesVersion, clusterObject.Spec.KubernetesVersion)
 
-	if etcdRestore.SnapshotRestore == kubernetesVersion || etcdRestore.SnapshotRestore == all {
-		clusterObject, _, err := clusters.GetProvisioningClusterByName(client, clusterName, namespace)
+		steveclient, err := client.Steve.ProxyDownstream(clusterID)
 		require.NoError(t, err)
-		require.Equal(t, initialKubernetesVersion, clusterObject.Spec.KubernetesVersion)
 
-		if strategy {
-			require.Equal(t, concurrencyDefaultValue, clusterObject.Spec.RKEConfig.UpgradeStrategy.ControlPlaneConcurrency)
-			require.Equal(t, concurrencyDefaultValue, clusterObject.Spec.RKEConfig.UpgradeStrategy.WorkerConcurrency)
+		deploymentList, err := steveclient.SteveType(workloads.DeploymentSteveType).NamespacedSteveClient(defaultNamespace).List(nil)
+		require.NoError(t, err)
+		require.Equal(t, 1, len(deploymentList.Data))
```

Comment:
does this fail if there were other jobs that left deployments in the default namespace? I would expect this to be able to still pass.

Reply:
I don't believe this will fail. At least, I haven't seen issues with this when doing multiple iterations across cluster types.